crates/accelerate/src/synthesis/clifford/random_clifford.rs

```diff
@@ -0,0 +1,163 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::synthesis::linear::utils::{
+    binary_matmul_inner, calc_inverse_matrix_inner, replace_row_inner, swap_rows_inner,
+};
+use ndarray::{concatenate, s, Array1, Array2, ArrayView2, ArrayViewMut2, Axis};
+use rand::{Rng, SeedableRng};
+use rand_pcg::Pcg64Mcg;
+
+/// Sample from the quantum Mallows distribution.
+fn sample_qmallows(n: usize, rng: &mut Pcg64Mcg) -> (Array1<bool>, Array1<usize>) {
+    // Hadamard layer
+    let mut had = Array1::from_shape_fn(n, |_| false);
+    // Permutation layer
+    let mut perm = Array1::from_shape_fn(n, |_| 0);
+    let mut inds: Vec<usize> = (0..n).collect();
+
+    for i in 0..n {
+        let m = n - i;
+        let eps: f64 = 4f64.powi(-(m as i32));
+        let r: f64 = rng.gen();
+        let index: usize = -((r + (1f64 - r) * eps).log2().ceil() as isize) as usize;
+        had[i] = index < m;
+        let k = if index < m { index } else { 2 * m - index - 1 };
+        perm[i] = inds[k];
+        inds.remove(k);
```

Comment:
Can this ever run into the problem that it's trying to access an entry that has been removed? Maybe it's obvious that it cannot but I don't quite see it directly 🤔

Reply:
Indeed, this was not clear to me either, but this is both a direct port of the now removed python function, and the pseudocode in the referenced paper.